polygeist/lib/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.cpp

```diff
@@ -169,17 +185,31 @@ void ReachingDefinitionAnalysis::visitOperation(
     if (isa<MemoryEffects::Read>(effect.getEffect()))
       continue;
 
-    // Update the reaching definition for the current value.
-    result |= after->setModifier(val, op);
+    // An allocate operation create a definition for the current value.
+    if (isa<MemoryEffects::Allocate>(effect.getEffect()))
+      result |= after->setModifier(val, op);
 
-    // Write operations might also update the reaching definitions of values
-    // that are aliased to the current one.
+    // A write operation updates the definition of the current value and the
+    // definition of its definitely aliased values. It also updates the
+    // potential definitions of values that may alias the current value.
     if (isa<MemoryEffects::Write>(effect.getEffect())) {
+      result |= after->setModifier(val, op);
       for (Value aliasedVal : aliasQueries.getMustAlias(val))
         result |= after->setModifier(aliasedVal, op);
       for (Value aliasedVal : aliasQueries.getMayAlias(val))
         result |= after->addPotentialModifier(aliasedVal, op);
     }
+
+    // A deallocate operation kills reaching definitions of the current value
+    // and of its definitely aliased values. It also kills the potential
+    // definitions of values that may alias the current value.
```

Comment:
Why should it kill the potential definitions?

Reply:
I think we should do `result |= after->removePotentialModifiers(val);`, but I am not sure about `result |= after->removePotentialModifiers(aliasedVal);`.